Initial copy functionality

nhinze23 opened the pull request:
Resolves #1

nhinze23 committed:
#1 Initial commit with copy functionality

nhinze23 committed:
#1 Resolve git root folder

cesmarvin commented on cmd/copier/main.go:
```diff
@@ -0,0 +1,79 @@
+
+func main() {
+	if len(os.Args) < 2 {
+		log.Fatal(fmt.Sprintf("expected at least on of the following commands: \n"+
```
🚫 **[checkstyle]** <gosimple> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>S1038: should use log.Fatalf(...) instead of log.Fatal(fmt.Sprintf(...))
<!-- __reviewdog__:ChA4Y2UzN2UwMzUyZGUxYThjEgpjaGVja3N0eWxl -->

cesmarvin commented on cmd/copier/main.go:
```diff
@@ -0,0 +1,79 @@
+	var targetPaths stringSliceFlag
+	copyCmd.Var(&sourcePaths, "source", "")
+	copyCmd.Var(&targetPaths, "target", "")
+	err := copyCmd.Parse(args)
```
🚫 **[checkstyle]** <ineffassign> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>ineffectual assignment to err
<!-- __reviewdog__:Cg80ZTczZTM0OGU2OTYzOTMSCmNoZWNrc3R5bGU= -->

cesmarvin commented on internal/copy/copier.go:
```diff
@@ -0,0 +1,52 @@
+		return fmt.Errorf("failed to flush buffer to file %s: %w", destFilePath, err)
+	}
+
+	log.Println(fmt.Sprintf("copied file %s to %s", srcfilePath, destFilePath))
```
🚫 **[checkstyle]** <gosimple> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>S1038: should use log.Printf(...) instead of log.Println(fmt.Sprintf(...))
<!-- __reviewdog__:ChAxYzY4Njk0ZDM4YjVhMjkzEgpjaGVja3N0eWxl -->

cesmarvin commented on internal/copy/volumeMount.go:
```diff
@@ -0,0 +1,141 @@
+	isSubPathMount := true
+
+	for src, dest := range srcToDest {
+		log.Println(fmt.Sprintf("Start copy files from dir %s to %s", src, dest))
```
🚫 **[checkstyle]** <gosimple> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>S1038: should use log.Printf(...) instead of log.Println(fmt.Sprintf(...))
<!-- __reviewdog__:ChA3ZTE5ZmUxNGZjM2VjYzc3EgpjaGVja3N0eWxl -->

cesmarvin commented on internal/copy/volumeMount.go:
```diff
@@ -0,0 +1,141 @@
+	for src, dest := range srcToDest {
+		log.Println(fmt.Sprintf("Start copy files from dir %s to %s", src, dest))
+		data := filepath.Join(src, "..data")
+		log.Println(fmt.Sprintf("Checking data symlink %s", data))
```
🚫 **[checkstyle]** <gosimple> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>S1038: should use log.Printf(...) instead of log.Println(fmt.Sprintf(...))
<!-- __reviewdog__:ChBiZTg1ODczYWY4OTBmZTMzEgpjaGVja3N0eWxl -->

nhinze23 committed:
#1 Fix sonar issues

cesmarvin commented on internal/copy/volumeMount.go:
```diff
@@ -0,0 +1,141 @@
+// the files are behind symlinks and the resolved folder is used as source. This path from src to the resolved folder
+// should not be copied to the destination.
+func (v *VolumeMountCopier) walk(srcVolume, destVolume, filePath string, isSubPathMount bool, d fs.DirEntry) error {
+	log.Println(fmt.Sprintf("Processing file %s", filePath))
```
🚫 **[checkstyle]** <gosimple> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>S1038: should use log.Printf(...) instead of log.Println(fmt.Sprintf(...))
<!-- __reviewdog__:ChBhNzIzNWRkYTQwNGQ4MWQ1EgpjaGVja3N0eWxl -->

cesmarvin commented on internal/copy/volumeMount.go:
```diff
@@ -0,0 +1,141 @@
+func (v *VolumeMountCopier) walk(srcVolume, destVolume, filePath string, isSubPathMount bool, d fs.DirEntry) error {
+	log.Println(fmt.Sprintf("Processing file %s", filePath))
+	if d.IsDir() {
+		log.Println(fmt.Sprintf("Skip dir %s", filePath))
```
🚫 **[checkstyle]** <gosimple> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>S1038: should use log.Printf(...) instead of log.Println(fmt.Sprintf(...))
<!-- __reviewdog__:ChA2MGI1MGVlZmQwYmYyOWNmEgpjaGVja3N0eWxl -->

cesmarvin commented on internal/copy/volumeMount.go:
```diff
@@ -0,0 +1,141 @@
+		}
+
+		if v.fileSystem.SameFile(sourceFileInfo, destFileInfo) {
+			log.Println(fmt.Sprintf("source file %s and destination file %s are equal", filePath, destinationFilePath))
```
🚫 **[checkstyle]** <gosimple> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>S1038: should use log.Printf(...) instead of log.Println(fmt.Sprintf(...))
<!-- __reviewdog__:ChBlOWI1OTYxMWJjY2Q1NjZjEgpjaGVja3N0eWxl -->

nhinze23 committed:
#1 Fix sonar issues

nhinze23 committed:
Merge remote-tracking branch 'origin/develop' into feature/1_initial_cp

# Conflicts:
#	build/make/bats/customBatsEntrypoint.sh
#	build/make/coder-lib.sh
#	build/make/prerelease.sh
#	build/make/release.sh
#	build/make/release_cve.sh
#	build/make/release_functions.sh
#	build/make/self-update.sh
#	build/make/trivyscan.sh
#	cmd/copier/main.go
#	go.mod
#	sonar-project.properties

nhinze23 committed:
#1 Exclude filesystem abstraction because unit test are not really possible for this file.

nhinze23 committed:
#1 Fix exclude

nhinze23 committed:
#1 Add coverage exclusion

meiserloh commented on cmd/copier/main_test.go:
```diff
@@ -0,0 +1,76 @@
+func Test_handleCopyCommand(t *testing.T) {
+	t.Run("should call copy subsequent src to destination volumes", func(t *testing.T) {
+		// given
+		oldCopyCmd := copyCmd
+		copyCmd = flag.NewFlagSet("copy", flag.ExitOnError)
```
Can you explain this to me? You save the old Cmd and then overwrite the copyCmd with the exact same code as in `main.go`, so this has no effect at all, does it?

meiserloh replied:
is this just preparation for future commands?

nhinze23 replied:
Oh you are right. This is useless. One can just reset the flagset in every test

meiserloh commented on docs/dev/usage_en.md:
```diff
@@ -0,0 +1,19 @@
+# Usage
+
+This application support currently following commands:
```
```suggestion
This application supports currently following commands:
```

meiserloh commented on docs/dev/usage_en.md:
```diff
@@ -0,0 +1,19 @@
+
+### Example
+
+`target/dogu-data-seeder copy -source=./cmd -target=./cmdCopy -source=./build -target=./buildCopy`
```
in your tests (`main_test.go`) you used two dashes for the args (so `--source=` instead of `-source=`). Do we have to adjust it here?

nhinze23 replied:
Both works. I will change it here for consistency.

meiserloh commented on docs/dev/usage_en.md:
```diff
@@ -0,0 +1,19 @@
+
+`target/dogu-data-seeder copy -source=./cmd -target=./cmdCopy -source=./build -target=./buildCopy`
+
+Where the source n will be copied to the target n.
```
```suggestion
Where each source will be copied to the immediately following target.
```

meiserloh commented on cmd/copier/main.go:
```diff
@@ -1,9 +1,82 @@
+		copyMap[sourcePaths[i]] = targetPaths[i]
+	}
+
+	err = copier.CopyVolumeMount(copyMap)
```
I was a little confused to not find the CopyVolumeMount in `copier.go`, because of the naming here. Maybe we can find a better one.

nhinze23 replied:
copier.go ist just for technically copy files. This can be used by other potential components in the future.

nhinze23 replied:
I renamed the vars to avoid confusion

meiserloh commented on internal/copy/volumeMount.go:
```diff
@@ -0,0 +1,141 @@
+	return &VolumeMountCopier{&fileSystem{}, copyFile}
+}
+
+// CopyVolumeMount copies all files from the given src path in srcToDest parameter to the given paths.
```
```suggestion
// CopyVolumeMount copies all files from the given src path in srcToDest parameter to the associate destination path.
```

meiserloh commented on internal/copy/volumeMount.go:
```diff
@@ -0,0 +1,141 @@
+}
+
+// CopyVolumeMount copies all files from the given src path in srcToDest parameter to the given paths.
+// It only handles regular files and stops if an error occurs.
```
This is confusing, since it will only stop the current copy and still continues with the next source-destination-pair

nhinze23 replied:
Oh yes this is old. The functionality now tries to copy as much as possible.

nhinze23 committed:
#1 Remove pkg docker copy because there are no contents

meiserloh commented on internal/copy/volumeMount_test.go:
```diff
@@ -0,0 +1,400 @@
+		require.NoError(t, err)
+	})
+
+	t.Run("should keep nested dirs from subpath mounts", func(t *testing.T) {
```
So this is cool, that this theoretically works, but in reality it would be hard for the operator to find out what to choose as a src-Dir in this case, wouldn't it?

nhinze23 replied:
Mh yes. A nested structure would consist of new mounts with a single dir. But let us keep this.

nhinze23 committed:
#1 Implement deletion of old unused files

nhinze23 committed:
#1 Add netrc for private dependencies

nhinze23 committed:
#1 don't panic

nhinze23 committed:
#1 Check file paths for existence on delete

nhinze23 committed:
#1 Always copy all volume mounts (by subpath and not) from the dir

nhinze23 committed:
#1 Use correct owner user and group

nhinze23 committed:
#1 Just reset flag set

nhinze23 committed:
#1 Apply review feedback for docs

nhinze23 committed:
#1 Rename vars to avoid confusion

nhinze23 committed:
#1 Fix spelling

nhinze23 committed:
#1 Delete obsolete comment

nhinze23 committed:
#1 Ensures no write permission to application in container

nhinze23 committed:
#1 Ensures no write permission to application in container

nhinze23 committed:
#1 Add template for netrc

nhinze23 committed:
#1 Update dependencies

nhinze23 committed:
#1 Fix execute permissions

nhinze23 committed:
#1 Skip walk on symlinks.

nhinze23 committed:
#1 Use correct config root dirs

nhinze23 committed:
#1 Only untrack files if they are really lost

nhinze23 committed:
#1 Delete vet analysis because no one will be generated

meiserloh commented on cmd/copier/main_test.go:
```diff
@@ -11,63 +11,110 @@ import (
 		require.NoError(t, err)
 	})
 
 	t.Run("should return error on copy error error", func(t *testing.T) {
```
```suggestion
	t.Run("should return error on copy error", func(t *testing.T) {
```

meiserloh commented on internal/copy/localConfig_test.go:
```diff
@@ -0,0 +1,221 @@
+  - /path/database
+  - /path/config
+`
+	expectedYamlFiles := `- /path/database
```
It's just a small thing, but format `actualYamlFiles` and `expectedYamlFiles` the same way, I just overread this line and was kind of confused

nhinze23 committed:
Update cmd/copier/main_test.go

Co-authored-by: meiserloh <132442180+meiserloh@users.noreply.github.com>

nhinze23 committed:
#1 Unify format of test artifacts

nhinze23 committed:
#1 Fix unit-test

nhinze23 committed:
#1 Set container user as the owner of the application.

Groups might changes depending of the volume group ids in the dogu json. This causes permission denied errors with fsGroup!=1000.
